File: src/shared/validators/user_role_validator.py
from src.modules.role.models import Role
from src.modules.role.domain.enums.role_references_enum import RoleReferences
from src.shared.types.user_request import UserRequest
from src.shared.exceptions.unauthorized_exception import UnauthorizedException
from src.shared.exceptions.not_found_exception import NotFoundException


def user_role_validator(role_id: int, user_request: UserRequest):

    if user_request is None:
        raise UnauthorizedException("Usuário não encontrado!")

    role = Role.objects.filter(id=role_id, deleted_at=None).first()

    if role is None:
        raise NotFoundException("Perfil não encontrado!")

    is_admin = user_request.role.reference == RoleReferences.ADMIN

    is_admin_organization = (
        user_request.role.reference == RoleReferences.ADMIN_ORGANIZATION
    )

    is_admin_company = user_request.role.reference == RoleReferences.ADMIN_COMPANY

    if role.reference == RoleReferences.ADMIN and is_admin:
        return

    if role.reference == RoleReferences.ADMIN_ORGANIZATION and is_admin:
        return

    if role.reference == RoleReferences.ADMIN_COMPANY and (
        is_admin or is_admin_organization
    ):
        return

    if role.reference == RoleReferences.CLIENT and (
        is_admin or is_admin_organization or is_admin_company
    ):
        return

    if user_request.role.id != role_id:
        raise UnauthorizedException("Usuário sem permissão nesse perfil!")

File: src/shared/validators/user-role.validator.ts
import { $Enums } from '@prisma/client';
import { UserRequest } from '../types/user-request';
import { PrismaService } from 'src/infra/database/prisma.service';
import { NotFoundException, UnauthorizedException } from '@nestjs/common';

export async function userRoleValidator(
  roleId: number,
  userRequest: UserRequest,
) {
  const prisma = new PrismaService();

  if (!userRequest) throw new UnauthorizedException('Usuário não encontrado!');

  const role = await prisma.role.findFirst({
    where: {
      id: roleId,
      deletedAt: null,
    },
  });

  const isAdmin: boolean =
    userRequest.role?.reference === $Enums.RoleReferences.ADMIN;

  const isAdminOrganization: boolean =
    userRequest.role?.reference === $Enums.RoleReferences.ADMIN_ORGANIZATION;

  const isAdminCompany: boolean =
    userRequest.role?.reference === $Enums.RoleReferences.ADMIN_COMPANY;

  if (!role) throw new NotFoundException('Perfil não encontrado!');

  if (role?.reference === $Enums.RoleReferences.ADMIN && isAdmin) return;

  if (role?.reference === $Enums.RoleReferences.ADMIN_ORGANIZATION && isAdmin)
    return;

  if (
    role?.reference === $Enums.RoleReferences.ADMIN_COMPANY &&
    (isAdmin || isAdminOrganization)
  )
    return;

  if (
    role?.reference === $Enums.RoleReferences.CLIENT &&
    (isAdmin || isAdminOrganization || isAdminCompany)
  )
    return;

  if (userRequest.roleId !== roleId)
    throw new UnauthorizedException('Usuário sem permissão nesse perfil!');
}
